opt: fix distinct-on negative limit hint

RaduBerinde opened the pull request:
The distinct on formula can return negative numbers in certain cases (leading to
a "negative limit hint" internal error); restrict its usage to the domain where
that doesn't happen and add a test.

Fixes #45571.
Fixes #45570.

Release note (bug fix): Fixed "negative limit hint" internal query error.

Release justification: Category 2 (bug fixes and low-risk updates to new functionality).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45879)
<!-- Reviewable:end -->

RaduBerinde commented:
Category 2: Bug fixes and low-risk updates to new functionality

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>  thanks!

Can you check whether this also fixes #44128 and the duplicate issues referenced in that issue?

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45879#-:-M1ztCYzB7CbepYEGHEX:bv6tx9r)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball and @RaduBerinde)

___
*[pkg/sql/opt/xform/physical_props_test.go, line 21 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45879#-M1zsimIE0cvVeo8kHQX:-M1zsimIE0cvVeo8kHQY:b4ivk4b) ([raw file](https://github.com/cockroachdb/cockroach/blob/433188c4d554cca10d46a5524b8a0757c668b22c/pkg/sql/opt/xform/physical_props_test.go#L21)):*
> ```Go
> 	var values []float64
> 	for _, v := range []float64{0, 1e-10, 1e-5, 0.1, 1, 10, 1e5, 1e10} {
> 		for _, n := range []float64{-1e10, +1e10, -1e7, +1e7} {
> ```

Can you add a comment about why you chose these values?


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
TFTR! Yeah they are the same issue.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45879#-:-M2-3Wav6ZrAVtyZ139W:beqzp79)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @rytaft)

___
*[pkg/sql/opt/xform/physical_props_test.go, line 21 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45879#-M1zsimIE0cvVeo8kHQX:-M2-3Zjn2obbtY8pGZTu:b-bky7bv) ([raw file](https://github.com/cockroachdb/cockroach/blob/433188c4d554cca10d46a5524b8a0757c668b22c/pkg/sql/opt/xform/physical_props_test.go#L21)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Can you add a comment about why you chose these values?
</blockquote></details>

These were meant to be `1e-10` etc. Fixed and added a comment, and improved the test to check for NaN and +Inf as well.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45879#-:-M20j_8lA0vWZUU-ICuK:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+

RaduBerinde committed:
opt: fix distinct-on negative limit hint

The distinct-on limit hint formula can return negative numbers in certain cases
(leading to a "negative limit hint" internal error); restrict its usage to the
domain where that doesn't happen and add a test.

Fixes #45571.
Fixes #45570.

Release note (bug fix): Fixed "negative limit hint" internal query error.

Release justification: Category 2 (bug fixes and low-risk updates to new functionality).

RaduBerinde commented:
bors r+

RaduBerinde commented:
bors r+
